src/symbol_table.cpp

```diff
@@ -0,0 +1,94 @@
+#include "symbol_table.hpp"
+
+#include "ast.hpp"
+#include "./log/log.hpp"
+
+#include <cassert>
+
+namespace Frontend {
+
+SymbolTable::SymbolTable() {
+	new_nested_scope();
+}
+
+SymbolTable::SymbolMap& SymbolTable::latest_shadowed_scope() {
+	return m_shadowed_scopes.back();
+}
+
+void SymbolTable::declare(AST::Declaration* decl) {
+	auto old_decl = [&]() -> AST::Declaration* {
+		auto insert_result = m_bindings.insert({decl->identifier_text(), decl});
+		if (insert_result.second) // introduced a previously undeclared name
+			return nullptr;
+		return std::exchange(insert_result.first->second, decl);
+	}();
+
+	auto insert_result =
+	    latest_shadowed_scope().insert({decl->identifier_text(), old_decl});
+	if (!insert_result.second)
+		Log::fatal() << "Redeclaration of '" << decl->identifier_text() << "'";
+}
```

Comment:
I need help to understand this

Reply:
you mean the construct itself? It is an immediately invoked lambda, it works not unlike a seq-expr in Jasper itself.

As for the logic of it: we try to insert a new declaration.

If we succeed, we know the name was previously undeclared, so there is no previous declaration, so we return nullptr.

If we fail, we get an iterator that points to the previous declaration, so we assign through that iterator, and return the old value. (which is what `std::exchange` does)

Here is a pseudocode for `std::exchange`
```c++
template<typename T>
T exchange(T& dst, T&& new_value) {
    T old_value = std::move(dst);
    dst = std::move(new_value);
    return old_value;
}
```

Reply:
Ahh, I see. Yea that makes sense.